src/components/core/GeoChart.tsx

```diff
@@ -74,44 +109,165 @@ export class GeoChartInner extends BaseVisualization<IGeoChartInnerProps, {}> {
+                    item.legendIndex,
+                    { ...item },
+                ),
+            };
+        });
+    };
+
+    private getSegmentBySeries(data: Execution.IResultHeaderItem[]): IPushpinCategoryLegendItem[] {
+        const segmentItemsUniq: Execution.IResultHeaderItem[] = unionBy(data, item => {
+            return getHeaderItemName(item);
+        });
+        const segmentItemNamesUniq: string[] = segmentItemsUniq.map(
+            (item: Execution.IResultHeaderItem): string => {
+                const name: string = getHeaderItemName(item);
+                return name || EMPTY_SEGMENT_ITEM;
+            },
+        );
+        const colorPaletteMapping = getColorPaletteMapping(segmentItemNamesUniq);
+        return segmentItemNamesUniq.map(
+            (segmentItemName: string, index: number): IPushpinCategoryLegendItem => {
+                const palette = colorPaletteMapping[segmentItemName];
+                const color = Array.isArray(palette) && palette.length > 0 ? palette[palette.length - 1] : "";
+                return {
+                    name: segmentItemName,
+                    legendIndex: index,
+                    color,
+                    isVisible: true,
+                };
+            },
+        );
+    }
+
+    private getGeoChartInnerOptions(props: IGeoChartInnerProps): IGeoChartInnerOptions {
+        if (!props.execution) {
+            return;
+        }
+        const {
+            config: { mdObject: { buckets = [] } = {} },
+            execution: { executionResult, executionResponse },
+            onDataTooLarge,
+        } = this.props;
+
+        const geoData: IGeoData = getGeoData(buckets, executionResponse.dimensions);
+        const validateResult = this.validateData(geoData);
+        if (!validateResult.dataTooLarge) {
+            // always force onDataTooLarge error handling
+            invariant(onDataTooLarge, "GeoChart's onDataTooLarge callback is missing.");
+            onDataTooLarge();
+        } else {
+            const { data } = executionResult;
+            const { size, segmentBy } = geoData;
+            const attributeHeaderItems = getGeoAttributeHeaderItems(executionResult, geoData);
+            let segmentByData: IPushpinCategoryLegendItem[] = [];
+            let sizeData: number[] = [];
+            if (size && isTwoDimensionsData(data)) {
+                sizeData = data[size.index].map(stringToFloat);
+            }
+            if (segmentBy) {
+                const segmentByAttributeHeaderItems = attributeHeaderItems[segmentBy.index];
+                segmentByData = this.getSegmentBySeries(segmentByAttributeHeaderItems);
+                if (this.state === undefined) {
+                    this.state = {
+                        legendItemsEnabled: [...segmentByData],
+                    };
+                } else if (!isEqual(this.state.legendItemsEnabled, segmentByData)) {
+                    this.setState({
```

Comment:
Please do `setState` outside `getGeoChartInnerOptions`.
Here we should do only get options.

Reply:
fixed